app/models/name.rb

```diff
@@ -408,33 +458,198 @@ class Name < AbstractModel
+          where(id: names.map(&:id)).with_correct_spelling
+        }
+  scope :text_name_includes,
+        ->(text_name) { where(Name[:text_name].matches("%#{text_name}%")) }
+  scope :with_classification, -> { where(Name[:classification].not_blank) }
+  scope :without_classification, -> { where(Name[:classification].blank) }
+  scope :classification_includes,
+        lambda { |classification|
+          where(Name[:classification].matches("%#{classification}%"))
+        }
+  scope :with_author, -> { where(Name[:author].not_blank) }
+  scope :without_author, -> { where(Name[:author].blank) }
+  scope :author_includes,
+        ->(author) { where(Name[:author].matches("%#{author}%")) }
+  scope :with_citation, -> { where(Name[:citation].not_blank) }
+  scope :without_citation, -> { where(Name[:citation].blank) }
+  scope :citation_includes,
+        ->(citation) { where(Name[:citation].matches("%#{citation}%")) }
+  scope :with_notes, -> { where(Name[:notes].not_blank) }
+  scope :without_notes, -> { where(Name[:notes].blank) }
+  scope :notes_include,
+        ->(notes) { where(Name[:notes].matches("%#{notes}%")) }
+  scope :with_comments, -> { joins(:comments).distinct }
+  scope :without_comments, -> { missing(:comments) }
+  scope :comments_include, lambda { |summary|
+    joins(:comments).where(Comment[:summary].matches("%#{summary}%")).distinct
+  }
+  scope :on_species_list,
+        lambda { |species_list|
+          joins(observations: :species_lists).
+            merge(SpeciesListObservation.where(species_list: species_list))
+        }
+  # Accepts region string, location_id, or Location instance
+  scope :at_location,
+        lambda { |location|
+          case location
+          when String # treat it as a region, not looking for all string matches
+            joins(observations: :location).
+              where(Location[:name].matches("%#{location}"))
+          when Integer, Location
+            joins(:observations).where(observations: { location: location })
+          else
+            Name.none
+          end
+        }
+  # Occurrence of name within a lat/long box
+  # TODO: Reexamine this:
+  #  As currently written, it finds Names whose Observations
+  #  are in Locations completely inside the box
+  #  And it's buggy (among other things, doesn't cover box straddling 180 def)
+  #  Probably better to just chain Observation.in_box(**args)
+  scope :in_box, # Use named parameters (n, s, e, w), any order
+        lambda { |**args|
+          if args[:s].present? && args[:n].present? &&
+             args[:w].present? && args[:e].present? &&
+             (args[:w] < args[:e])
+
+            joins(observations: :location).where(
+              (Location[:south] >= args[:s]).and(Location[:north] <= args[:n]).
+              and(Location[:west] >= args[:w]).and(Location[:east] <= args[:e]).
+              and(Location[:west] <= Location[:east])
+            )
+          else
+            Name.none
+          end
+        }
```

Comment:
@nimmolo, @pellaea 
I'm looking for some advice about the intent and use of `Name scope :in_box`
It should be tested and re-written, but exactly how depends on intent.
- Is it intended to replace a current query or search? (I couldn't find one.)
- Should it depend on:
  - The Locations of Observations of Name vs
  - The lat/lon of Observations of Names

It could/should be refactored and tested as something like (quick & dirty guess, you'll get the idea)
```ruby
joins(observations: :locations).merge(Location.in_box(args))
```
or
```ruby
joins(observations).merge(Observation..in_box(args))
```

Each of these has its advantages and flaws.
Maybe I should delete it until we have a better idea about how it will be used.

Reply:
PS: Here's the commit message from when it was added
>Add Name.in_box(n, s, e, w), tidy up, add TODOs
Commit:
3d057c8400ee24b72152fa779aae555a4b2ca4c0 [3d057c840]
Parents:
[77354dc802](rev://77354dc802deb06eaa8224af6978be56d31fd970)
Author:
andrew nimmo <andrnimm@fastmail.fm>
Date:
July 20, 2022 at 2:31:58 PM PDT

Reply:
This can actually be done on the site now by doing a search for observations with north, south, east, west parameters, and then converting the search to names.  It kind of makes sense that people might want to do this.  It comes up with a list of species that have been observed in a given region.

However, I would personally like to avoid two- or more-level joins.  I know Query doesn't currently work that way, but I would like for it to some day.  In this particular case, I think the ideal would be to write a scope for Observation, and then do this query as `select from names join observations where observations.id in (subquery)`, where the subquery is the Observation.in_box scope.  I did benchmarks on this a few years ago, and it was significantly faster.  I bet it's only gotten better as the database has grown.

This can be converted to a scope, too, if you must, although I don't remember the syntax Nimmo figured out for the subquery.  It is something `Name.joins(:observations).where(... observations.id in Observation.in_box(params) ...)`.  I wonder if we could make instead a generic Name scope that takes as an argument the observation scope.  That would save a huge amount of effort!

I think a good general pattern for the UI would be to focus on observations.  Write all of the scopes for observations.  Then let the user request the locations, names, species lists, sequences, images, or whatever that are associated with those observations.  It would get rid of all of the silly Query "coercion" code completely.  It would ensure all of the actual SQL queries are reasonable.  It keeps all of the scopes in one place (Observation). When we provide the results for the associated objects the queries can all be implemented using the subquery technique above.

Compare to the way Query works now, where each of the (many) observation scopes is essentially duplicated in names, locations and images.  (And should be duplicated in all the other associated classes, too.)  And furthermore, consider how heinous the performance would be for an Images.in_box scope(!) Joining from images through observation_images through observations to locations is horrendous.  Done instead as a subquery, it's actually reasonably efficient.

Does this make sense?

Reply:
@pellaea & @nimmolo Thanks you both for the replies.
TLDR: I'll implement Name scope: in_box as `joins(:observations).merge(Observations.in_box(args))` (or something close. I often have trouble remembering the right syntax).

Other:
- If we don't like this, we can fix it.
- I might not get to this until Sat. I'm about to head to bed, and won't have electricity most of Fri.
- @nimmolo: I checked PatternSearchTest but couldn't find anything like this. 
- There's already a Name scope `:at_location`, so `in_box` must refer to something else. The analogy is Observation scope `:in_box`.
- As usual, you guys are thinking about straightforward, useful things, while I've (almost) immediately jumped to edge cases
- An Observation can have 3 location-y attributes: 
  - a point defined by `lat`/`long`; 
  - a box defined by`location`; 
  - a purely textual `where`
- Sometimes these don't match up. For instance:
  - Many Observations have Locations, but no lat/lon. 
  - Many Observations do not use the smallest possible Location. E.g., instead of a city, they have a county, state, country, or event Earth.
- Some of the shortcomings of various scopes
  - `Observation.at_location(loc)` will miss Observations whose lat//lon is within the bounds of `loc`, but use a different location -- whether bigger, smaller, or even identical boundaries but a different Location.name.
  - `Observation in_box(box)` will miss Observations that lack lat/lon (typical of older Observations), but whose Location is completely within `box`.
- A Name scope will inherit the shortcomings of chained Observation scopes.
- I therefore thought -- based on the first draft of `Name.in_box` -- that it was intended to be something different. Names of Observations whose Location is completely in `box`.  That will catch some of the cases missed by `Name.at_loc`.
- If we want to be more complete, we'd have something like `:in_box_or_location_in_box`. 
It would look for Names.in_box or Names whose location is completely in the box.

Reply:
I think this is also a case of you trying to do something too perfect with too imperfect data!

I agree that whatever shortcomings are present in Observation.in_box will also show here.  I think, however, instead of working on it here, it should be fixed in Observation.  To the extent that we remove all names and locations without observations, there is nothing to be gained by working on the name or location version of any scope.  Again, I think it is worth focusing on the observation, and then doing the name and location scopes as just join + merge with the corresponding observation scope.

I don't know how we fix all edge cases you're thinking of.

Do we trust the observation lat/long more than the location box, which we trust more than the text location?  If so, it suggests at least an order: first use the obs lat/long if present (and ignore the location box and text location), or if the obs lat/long is absent, try the location box, and failing that fall back on the text location (if that's meaningful: if the user gives us a lat/long box, I don't see how we can do anything with the text location.)

Reply:
@pellaea I agree, and also think that `joins(:observations).merge(Observations.in_box(args))` is good enough for now.
If we want to fine tune it, we can do that later.